watchtower/wtdb/migration8/migration.go

```diff
@@ -0,0 +1,224 @@
+	//    seqnum -> encoded CommittedUpdate.
+	cSessionCommits = []byte("client-session-commits")
+
+	// cChanDetailsBkt is a top-level bucket storing:
+	//   channel-id => cChannelSummary -> encoded ClientChanSummary.
+	//  		=> cChanDBID -> db-assigned-id
+	// 		=> cChanSessions => db-session-id -> 1
+	// 		=> cChanClosedHeight -> block-height
+	// 		=> cChanMaxCommitmentHeight -> commitment-height
+	cChanDetailsBkt = []byte("client-channel-detail-bucket")
+
+	cChanMaxCommitmentHeight = []byte(
+		"client-channel-max-commitment-height",
+	)
+
+	// ErrUninitializedDB signals that top-level buckets for the database
+	// have not been initialized.
+	ErrUninitializedDB = errors.New("db not initialized")
+
+	byteOrder = binary.BigEndian
+)
+
+// MigrateChannelMaxHeights migrates the tower client db by collecting all the
+// max commitment heights that have been backed up for each channel and then
+// storing those heights alongside the channel info.
+func MigrateChannelMaxHeights(tx kvdb.RwTx) error {
+	log.Infof("Migrating the tower client DB for quick channel max " +
+		"commitment height lookup")
+
+	heights, err := collectChanMaxHeights(tx)
+	if err != nil {
+		return err
+	}
+
+	return writeChanMaxHeights(tx, heights)
+}
+
+// writeChanMaxHeights iterates over the given channel ID to height map and
+// writes an entry under the cChanMaxCommitmentHeight key for each channel.
+func writeChanMaxHeights(tx kvdb.RwTx, heights map[ChannelID]uint64) error {
+	chanDetailsBkt := tx.ReadWriteBucket(cChanDetailsBkt)
+	if chanDetailsBkt == nil {
+		return ErrUninitializedDB
+	}
+
+	for chanID, maxHeight := range heights {
+		chanDetails := chanDetailsBkt.NestedReadWriteBucket(chanID[:])
+
+		// If the details bucket for this channel ID does not exist,
+		// it is probably a channel that has been closed and deleted
+		// already. So we can skip this height.
+		if chanDetails == nil {
+			continue
+		}
+
+		b, err := writeBigSize(maxHeight)
+		if err != nil {
+			return err
+		}
+
+		err = chanDetails.Put(cChanMaxCommitmentHeight, b)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// collectChanMaxHeights iterates over all the sessions in the DB. For each
+// session, it iterates over all the Acked updates and the committed updates
+// to collect the maximum commitment height for each channel.
+func collectChanMaxHeights(tx kvdb.RwTx) (map[ChannelID]uint64, error) {
+	sessionsBkt := tx.ReadBucket(cSessionBkt)
+	if sessionsBkt == nil {
+		return nil, ErrUninitializedDB
+	}
+
+	chanIDIndexBkt := tx.ReadBucket(cChanIDIndexBkt)
+	if chanIDIndexBkt == nil {
+		return nil, ErrUninitializedDB
+	}
+
+	heights := make(map[ChannelID]uint64)
+
+	// For each update we consider, we will only update the heights map if
+	// the commitment height for the channel is larger than the current
+	// max height stored for the channel.
+	cb := func(chanID ChannelID, commitHeight uint64) {
+		height, ok := heights[chanID]
```

Comment:
nit: there's no need for the `ok` since it'll just return zero if the key is not present. So could just do `if commitHeight > heights[chanID]`...

Reply:
indeed! thanks!